File: .github/automation/issue-85-room-status-facet-badges.mjs
import assert from 'node:assert/strict';
import { chromium } from 'playwright';

const base='http://127.0.0.1:4173/index.html#scenario=0&role=admin&view=rooms&date=2026-08-15&filter=all&type=all&q=';
const errors=[];
const browser=await chromium.launch({headless:true});
function observe(page,label){page.on('pageerror',error=>errors.push(`${label} pageerror: ${error.message}`));page.on('console',message=>{if(message.type()==='error')errors.push(`${label} console: ${message.text()}`);});}
async function api(page,method,...args){return page.evaluate(({method,args})=>{const fn=window.__CASTLE_TEST__?.[method];if(typeof fn!=='function')throw new Error(`test API missing: ${method}`);return fn(...args);},{method,args});}
async function waitForFacetCard(page,room){await page.waitForFunction(room=>Array.from(document.querySelectorAll('.room-card-v2,.room-card')).some(card=>Array.from(card.querySelectorAll('[data-id]')).some(node=>node.dataset.id===room)&&card.querySelector('.room-concurrent-statuses')),room);}

const page=await browser.newPage({viewport:{width:1440,height:1000}});observe(page,'desktop');await page.goto(base,{waitUntil:'networkidle'});await api(page,'resetScenario',0);
const durableStart=await api(page,'fingerprint');

// Occupied + stayover cleaning + operation stopped.
let candidates=await api(page,'manualCleaningCandidates');
const occupied=String(candidates.find(item=>item.occupancy==='occupied')?.room||'');
assert.ok(occupied,'occupied test candidate missing');
await api(page,'setManualCleaning',occupied,true);
let facets=await api(page,'setRoomStatusFacetTest',occupied,{roomStopped:true,stopReason:'시설 안전 확인'});
assert.equal(facets.occupied,true);assert.equal(facets.cleaningActive,true);assert.equal(facets.operationStopped,true);
await api(page,'setRoomFilter','cleaning');await waitForFacetCard(page,occupied);
let dom=await api(page,'concurrentStatusDom',occupied);
assert.match(dom.cardText,/투숙 중/);assert.match(dom.cardText,/연박 청소/);assert.match(dom.cardText,/운영 중지/);
await api(page,'showRoom',occupied);await page.waitForFunction(room=>!!document.querySelector(`[data-room-state-dimensions="${room}"]`),occupied);
dom=await api(page,'concurrentStatusDom',occupied);assert.match(dom.panelText,/투숙 중/);assert.match(dom.panelText,/연박 청소/);assert.match(dom.panelText,/운영 중지/);

// Checked-out + checkout inspection + checkout cleaning.
await api(page,'resetScenario',0);await api(page,'setOperationalMoment','2026-08-15','13:05');
const momentCandidates=await api(page,'concurrentStatusCandidates');
const checkoutCase=momentCandidates.find(item=>item.checkoutInspection);
assert.ok(checkoutCase,'checkout-inspection test candidate missing at 13:05');
const checkoutRoom=String(checkoutCase.room);
assert.equal(checkoutCase.occupied,false);assert.equal(checkoutCase.cleaningActive,true,'checkout inspection room should retain checkout cleaning flow');
const checkoutFilter=await api(page,'setRoomFilter','checkout-inspection');assert.ok(checkoutFilter.includes(checkoutRoom));await waitForFacetCard(page,checkoutRoom);
dom=await api(page,'concurrentStatusDom',checkoutRoom);
assert.match(dom.cardText,/공실/);assert.match(dom.cardText,/퇴실점검 대상/);assert.match(dom.cardText,/퇴실 청소|청소/);
await api(page,'showRoom',checkoutRoom);await page.waitForFunction(room=>!!document.querySelector(`[data-room-state-dimensions="${room}"]`),checkoutRoom);
dom=await api(page,'concurrentStatusDom',checkoutRoom);assert.match(dom.panelText,/퇴실점검/);assert.match(dom.panelText,/청소/);

// Vacant + additional cleaning + candles.
await api(page,'resetScenario',0);candidates=await api(page,'manualCleaningCandidates');
const vacant=String(candidates.find(item=>item.occupancy!=='occupied')?.room||'');assert.ok(vacant,'vacant test candidate missing');
await api(page,'setManualCleaning',vacant,true);facets=await api(page,'setRoomStatusFacetTest',vacant,{candles:2});
assert.equal(facets.occupied,false);assert.equal(facets.cleaningActive,true);assert.equal(facets.candleCount,2);
await api(page,'setRoomFilter','cleaning');await waitForFacetCard(page,vacant);dom=await api(page,'concurrentStatusDom',vacant);
assert.match(dom.cardText,/공실/);assert.match(dom.cardText,/추가 청소/);assert.match(dom.cardText,/촛불 2개/);

// Data hold is a separate facet and does not erase occupancy.
await api(page,'resetScenario',0);candidates=await api(page,'manualCleaningCandidates');
const dataRoom=String(candidates.find(item=>item.occupancy==='occupied')?.room||'');assert.ok(dataRoom);
facets=await api(page,'setRoomStatusFacetTest',dataRoom,{dataIssue:'객실 기준정보 확인 필요'});
assert.equal(facets.occupied,true);assert.equal(facets.dataHold,true);
await api(page,'setRoomFilter','occupied');await waitForFacetCard(page,dataRoom);dom=await api(page,'concurrentStatusDom',dataRoom);assert.match(dom.cardText,/투숙 중/);assert.match(dom.cardText,/정보 확인 필요/);
await api(page,'setRoomStatusFacetTest',dataRoom,{dataIssue:''});

const repeat=await api(page,'repeatRender',10);assert.equal(repeat.equal,true,'re-rendering changed durable ledgers');await api(page,'assertUnique');
const duplicateRows=await page.evaluate(()=>Array.from(document.querySelectorAll('.room-card-v2,.room-card')).filter(card=>card.querySelectorAll(':scope > .room-concurrent-statuses').length>1).length);assert.equal(duplicateRows,0,'duplicate facet rows were inserted');

for(const width of [390,768,1440]){
  const responsive=await browser.newPage({viewport:{width,height:1000}});observe(responsive,`responsive-${width}`);await responsive.goto(base,{waitUntil:'domcontentloaded'});const list=await api(responsive,'manualCleaningCandidates');const room=String(list.find(item=>item.occupancy==='occupied')?.room||'');assert.ok(room);await api(responsive,'setManualCleaning',room,true);await api(responsive,'setRoomStatusFacetTest',room,{roomStopped:true,stopReason:'반응형 테스트'});await api(responsive,'setRoomFilter','cleaning');await waitForFacetCard(responsive,room);const text=await api(responsive,'concurrentStatusDom',room);assert.match(text.cardText,/투숙 중/);assert.match(text.cardText,/연박 청소/);assert.match(text.cardText,/운영 중지/);const overflow=await responsive.evaluate(()=>document.documentElement.scrollWidth-window.innerWidth);assert.ok(overflow<=1,`${width}px horizontal overflow: ${overflow}`);await responsive.close();
}

const durableEnd=await api(page,'fingerprint');assert.ok(durableStart&&durableEnd);await page.close();await browser.close();assert.deepEqual(errors,[],errors.join('\n'));console.log('Rendered room status facet combinations and responsive QA: passed');

File: .github/automation/issue-85-room-status-facet-badges.py
from __future__ import annotations

import hashlib
import json
from datetime import datetime
from pathlib import Path
from zoneinfo import ZoneInfo

HTML=Path('WIREFRAME/index.html')
html=HTML.read_text(encoding='utf-8')


def replace_between(start_marker:str,end_marker:str,replacement:str,label:str)->None:
    global html
    start=html.find(start_marker)
    if start<0:raise SystemExit(f'{label}: start marker missing')
    end=html.find(end_marker,start)
    if end<0:raise SystemExit(f'{label}: end marker missing')
    html=html[:start]+replacement+html[end:]


def replace_once(old:str,new:str,label:str)->None:
    global html
    count=html.count(old)
    if count!=1:raise SystemExit(f'{label}: expected one match, found {count}')
    html=html.replace(old,new,1)

facets=r'''      function roomStatusFacets(no) {
        no=String(no);
        const room=ROOMS.find(item=>item.no===no),cleaning=roomCleaningFacet(no),occupied=room?.occupancy==='occupied',blockingReasons=room?roomBlockingReasons(no):[],operationStopped=!!state.roomStopped?.[no],dataHold=!!room&&roomIsOnHold(no),candleCount=room&&!occupied?Number(state.candles?.[no]||0):0,checkoutInspection=!!checkoutInspectionPending(no);
        const genericBlockingReasons=blockingReasons.filter(reason=>!String(reason).startsWith('운영 중지')&&!String(reason).startsWith('정보 확인 필요')&&!String(reason).startsWith('촛불'));
        const genericBlocked=genericBlockingReasons.length>0;
        return {
          room:no,
          occupied,
          occupancy:occupied?'occupied':'vacant',
          occupancyLabel:occupied?'투숙 중':'공실',
          cleaningActive:cleaning.active,
          cleaningKind:cleaning.kind,
          cleaningStage:cleaning.stage,
          cleaningLabel:cleaning.label,
          cleaningTone:cleaning.tone,
          cleaningJob:cleaning.job,
          checkoutInspection,
          operationStopped,
          dataHold,
          genericBlocked,
          blockingReasons,
          genericBlockingReasons,
          candleCount,
          blocked:operationStopped||dataHold||genericBlocked||candleCount>0,
          hasSecondary:cleaning.active||checkoutInspection||operationStopped||dataHold||genericBlocked||candleCount>0,
        };
      }
'''
replace_between('      function roomStatusFacets(no) {','      function roomMatchesFacetFilter',facets,'room status facets')

badges_and_panel=r'''      function roomFacetBadgeMarkup(facets) {
        if(!facets.hasSecondary)return '';
        const badges=[statusBadge(facets.occupancyLabel,facets.occupied?'blue':'neutral')];
        if(facets.cleaningActive)badges.push(statusBadge(facets.cleaningLabel||`${facets.cleaningKind||'청소'} 필요`,facets.cleaningTone));
        if(facets.checkoutInspection)badges.push(statusBadge('퇴실점검 대상','amber'));
        if(facets.operationStopped)badges.push(statusBadge('운영 중지','red'));
        else if(facets.dataHold)badges.push(statusBadge('정보 확인 필요','amber'));
        if(facets.genericBlocked)badges.push(statusBadge('입실 차단','red'));
        if(facets.candleCount>0)badges.push(statusBadge(`촛불 ${facets.candleCount}개`,'amber'));
        return badges.join('');
      }
      function concurrentStatusBadges(no) { return roomFacetBadgeMarkup(roomStatusFacets(no)); }
      function concurrentStatusPanelMarkup(no) {
        const facets=roomStatusFacets(no);if(!facets.hasSecondary)return '';
        const items=[`<div class="info-item"><span>점유</span><strong>${esc(facets.occupancyLabel)}</strong></div>`];
        if(facets.cleaningActive){items.push(`<div class="info-item"><span>청소 유형</span><strong>${esc(facets.cleaningKind||'청소')}</strong></div>`);items.push(`<div class="info-item"><span>청소 단계</span><strong>${esc(facets.cleaningStage||'청소 필요')}</strong></div>`);}
        if(facets.checkoutInspection)items.push('<div class="info-item"><span>퇴실점검</span><strong>대상 · 완료 전</strong></div>');
        if(facets.operationStopped)items.push(`<div class="info-item"><span>운영</span><strong>운영 중지</strong></div>`);
        else if(facets.dataHold)items.push('<div class="info-item"><span>기준정보</span><strong>정보 확인 필요</strong></div>');
        if(facets.genericBlocked)items.push(`<div class="info-item"><span>입실 조건</span><strong>${esc(facets.genericBlockingReasons.join(' · ')||'입실 차단')}</strong></div>`);
        if(facets.candleCount>0)items.push(`<div class="info-item"><span>촛불</span><strong>${facets.candleCount}개 회수 필요</strong></div>`);
        return `<section class="card card-pad room-state-dimensions" data-room-state-dimensions="${no}" aria-labelledby="room-state-dimensions-title-${no}"><div class="section-head"><div><h3 id="room-state-dimensions-title-${no}">동시에 적용되는 객실 상태</h3><p class="audit-note">점유·청소·퇴실점검·운영·안전 조건을 서로 덮어쓰지 않고 함께 표시합니다.</p></div></div><div class="badge-row">${roomFacetBadgeMarkup(facets)}</div><div class="info-grid">${items.join('')}</div></section>`;
      }
'''
replace_between('      function concurrentStatusBadges(no) {','      function decorateConcurrentRoomStatuses',badges_and_panel,'status badge and detail panel')

replace_once(
"""            card.dataset.occupancyFacet=facets.occupancy;
            card.dataset.cleaningFacet=facets.cleaningActive?'active':'none';
            card.dataset.cleaningKind=facets.cleaningKind||'';""",
"""            card.dataset.occupancyFacet=facets.occupancy;
            card.dataset.cleaningFacet=facets.cleaningActive?'active':'none';
            card.dataset.cleaningKind=facets.cleaningKind||'';
            card.dataset.checkoutInspectionFacet=facets.checkoutInspection?'pending':'none';
            card.dataset.operationFacet=facets.operationStopped?'stopped':facets.dataHold?'data-hold':facets.genericBlocked?'blocked':'normal';
            card.dataset.candleFacet=String(facets.candleCount||0);""",
'card facet datasets')

api=r'''          setRoomStatusFacetTest:(roomNo,patch={})=>{const no=String(roomNo),room=ROOMS.find(item=>item.no===no);if(!room)throw new Error('객실을 찾을 수 없습니다.');if(Object.prototype.hasOwnProperty.call(patch,'roomStopped')){state.roomStopped[no]=!!patch.roomStopped;if(patch.roomStopped)state.roomStopReasons[no]=String(patch.stopReason||'상태 조합 테스트');else delete state.roomStopReasons[no];}if(Object.prototype.hasOwnProperty.call(patch,'dataIssue'))room.dataIssue=String(patch.dataIssue||'');if(Object.prototype.hasOwnProperty.call(patch,'candles'))state.candles[no]=Math.max(0,Number(patch.candles)||0);render();return {...roomStatusFacets(no)};},
'''
marker='          roomStatusFacets:roomNo=>({...roomStatusFacets(String(roomNo))}),'
if html.count(marker)!=1:raise SystemExit(f'test API marker mismatch: {html.count(marker)}')
html=html.replace(marker,marker+'\n'+api,1)

HTML.write_text(html,encoding='utf-8')

Path('DOCS/22_ROOM_STATUS_FACET_BADGES.md').write_text('''# 객실 상태 차원 배지 정책\n\n- 확정일: 2026-08-24\n\n객실 카드의 단일 주 상태는 요약일 뿐이며, 중요한 운영 조건을 덮어쓰지 않는다. 청소·퇴실점검·운영 중지·정보 확인·입실 차단·촛불 중 하나라도 활성화되면 카드에 점유 상태와 모든 보조 상태를 함께 표시한다.\n\n## 표시 순서\n\n1. 점유: 투숙 중 / 공실\n2. 청소: 유형과 현재 단계\n3. 퇴실점검 대상\n4. 운영 중지 또는 정보 확인 필요\n5. 기타 입실 차단\n6. 촛불 회수 수량\n\n객실 상세의 `동시에 적용되는 객실 상태` 패널도 같은 차원을 사용한다. 상태 필터는 각 차원을 독립적으로 검색한다.\n''',encoding='utf-8')
for path,appendix in [
('WIREFRAME/README.md','''\n\n## 전체 객실 상태 차원 배지 (2026-08-24)\n\n청소·퇴실점검·운영·안전 보조 조건이 있는 객실은 점유 상태와 모든 활성 차원을 카드와 상세에 함께 표시한다.\n'''),
('WIREFRAME/QA.md','''\n\n## 2026-08-24 · 전체 객실 상태 차원 조합\n\n- 투숙+연박 청소+운영 중지, 공실+퇴실점검+퇴실 청소, 공실+추가 청소+촛불 조합을 확인했다.\n- 각 상태가 카드·상세에서 동시에 보이고 기존 필터가 유지되는지 확인했다.\n- 배지 중복, 반복 렌더링 원장 변경, 390·768·1440px 가로 넘침과 콘솔 오류를 확인했다.\n''')]:
    file=Path(path);file.write_text(file.read_text(encoding='utf-8').rstrip()+appendix,encoding='utf-8')

checker=Path('scripts/check-workspace.mjs')
text=checker.read_text(encoding='utf-8')
marker="console.log('Workspace check: passed');"
contracts=r'''for (const contract of [
  'function roomFacetBadgeMarkup(facets)',
  "statusBadge('퇴실점검 대상','amber')",
  "statusBadge('운영 중지','red')",
  "statusBadge('정보 확인 필요','amber')",
  "statusBadge('입실 차단','red')",
  '점유·청소·퇴실점검·운영·안전 조건을 서로 덮어쓰지 않고 함께 표시합니다.',
  'data-checkout-inspection-facet',
  'setRoomStatusFacetTest:',
]) {
  if (!html.includes(contract)) throw new Error(`Room status facet badge contract missing: ${contract}`);
}
console.log('Room status facet badge static contracts: passed');

'''
if text.count(marker)!=1:raise SystemExit(f'checker marker mismatch: {text.count(marker)}')
checker.write_text(text.replace(marker,contracts+marker,1),encoding='utf-8')

sha=hashlib.sha256(HTML.read_bytes()).hexdigest()
sums=Path('SHA256SUMS.txt');lines=sums.read_text(encoding='utf-8').splitlines();found=False;out=[]
for line in lines:
    if line.endswith('  WIREFRAME/index.html'):out.append(f'{sha}  WIREFRAME/index.html');found=True
    else:out.append(line)
if not found:raise SystemExit('wireframe checksum missing')
sums.write_text('\n'.join(out)+'\n',encoding='utf-8')
manifest_path=Path('manifest.json');manifest=json.loads(manifest_path.read_text(encoding='utf-8'));manifest['version']='2026-08-24-room-status-facet-badges';manifest['generated_at_kst']=datetime.now(ZoneInfo('Asia/Seoul')).isoformat(timespec='seconds');manifest.setdefault('sha256',{})['WIREFRAME/index.html']=sha;manifest_path.write_text(json.dumps(manifest,ensure_ascii=False,indent=2)+'\n',encoding='utf-8')
